packages/docusaurus-plugin-sitemap/src/createSitemap.ts

```diff
@@ -75,18 +78,30 @@ export default async function createSitemap(
     return null;
   }
 
-  const sitemapStream = new SitemapStream({hostname});
+  function mapRoute(route: RouteConfig) : RRRouteConfig {
+    const { component, routes: originalRoutes, ...rrRoute } = route;
+    return { routes: originalRoutes?.map(mapRoute), ...rrRoute };
+  }
+
+  const sitemapStream = new SitemapStream({hostname, lastmodDateOnly: options.lastmod === 'date'});
+  const rrRoutes: RRRouteConfig[] = routes.map(mapRoute);
 
-  includedRoutes.forEach((routePath) =>
+  includedRoutes.forEach((routePath) => {
     sitemapStream.write({
       url: applyTrailingSlash(routePath, {
         trailingSlash: siteConfig.trailingSlash,
         baseUrl: siteConfig.baseUrl,
       }),
       changefreq,
       priority,
-    }),
-  );
+      lastmod:
+        routes !== undefined && options.lastmod !== false
+          ? matchRoutes(rrRoutes, routePath).find(
+            ({route}) => route?.exact === true,
+          )?.route?.lastModified
+          : undefined,
+    });
+  });
```

Comment:
that's not ideal to use `matchRoute` this way. 

`matchRoute` can be expensive particularly for very large sites with thousands of routes

We should create the sitemap directly from the new `routes` attribute.

We have an util somewhere in core to do exactly that, and is what we use to create the `routesPaths` attribute.

```js
export function getAllFinalRoutes(routeConfig: RouteConfig[]): RouteConfig[] {
  function getFinalRoutes(route: RouteConfig): RouteConfig[] {
    return route.routes ? route.routes.flatMap(getFinalRoutes) : [route];
  }
  return routeConfig.flatMap(getFinalRoutes);
}
```

Honestly I'm not a fan of the `routesPaths` attribute, and will likely try to remove it in favor of just `routes` in the future, so I'll take this opportunity to refactor the sitemap plugin to not use `routesPaths` and use `routes` instead.